lib/Dialect/Triton/Transforms/RewriteTMAToLegacy.cpp

```diff
@@ -0,0 +1,428 @@
+  TypedAttr attr = builder.getZeroAttr(elementType);
+
+  // Float NaN padding case
+  if (padding == triton::PaddingOption::PAD_NAN) {
+    assert(!elementType.isIntOrIndex());
+    auto apNaN =
+        llvm::APFloat::getNaN(cast<FloatAttr>(attr).getValue().getSemantics());
+    attr = builder.getFloatAttr(elementType, apNaN);
+  }
+
+  // Create tensor
+  Value constant = builder.create<arith::ConstantOp>(loc, attr);
+  return builder.create<triton::SplatOp>(loc, otherTensorType, constant);
+}
+
+// Note the attributes below don't normally appear on the TMA load/store ops.
+// However, if we are lowering block pointers these will be added as addition
+// attributes on the TMA operation for us to retrieve.
+std::optional<ArrayRef<int>> getBoundaryCheck(Operation *op) {
+  std::string_view attr_name = "boundaryCheck";
+  if (auto attr = op->getAttrOfType<DenseI32ArrayAttr>(attr_name)) {
+    return attr.asArrayRef();
+  }
+  return std::nullopt;
+}
+std::optional<triton::PaddingOption> getPadding(Operation *op) {
+  std::string_view attr_name = "padding";
+  if (auto attr = op->getAttrOfType<triton::PaddingOptionAttr>(attr_name)) {
+    return attr.getValue();
+  }
+  return std::nullopt;
+}
+triton::CacheModifier getCache(Operation *op) {
+  std::string_view attr_name = "cache";
+  if (auto attr = op->getAttrOfType<triton::CacheModifierAttr>(attr_name)) {
+    return attr.getValue();
+  }
+  return triton::CacheModifier::NONE;
+}
+triton::EvictionPolicy getEvict(Operation *op) {
+  std::string_view attr_name = "evict";
+  if (auto attr = op->getAttrOfType<triton::EvictionPolicyAttr>(attr_name)) {
+    return attr.getValue();
+  }
+  return triton::EvictionPolicy::NORMAL;
+}
+bool getIsVolatile(Operation *op) {
+  std::string_view attr_name = "isVolatile";
+  if (auto attr = op->getAttrOfType<BoolAttr>(attr_name)) {
+    return attr.getValue();
+  }
+  return false;
+}
+
+SmallVector<mlir::Value> castToI64(OpBuilder &builder,
+                                   mlir::ValueRange values) {
+  auto i64_type = builder.getI64Type();
+  return llvm::map_to_vector(values, [&](mlir::Value v) {
+    return builder.createOrFold<arith::ExtSIOp>(v.getLoc(), i64_type, v);
+  });
+}
+
+struct RewriteMakeTensorDesc : OpConversionPattern<triton::MakeTensorDescOp> {
+  using OpConversionPattern<triton::MakeTensorDescOp>::OpConversionPattern;
+
+  llvm::LogicalResult
+  matchAndRewrite(triton::MakeTensorDescOp op, OneToNOpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    SmallVector<mlir::Value> ptr_shape_strides;
+    // Note that none of these values come from a tensor descriptor so its safe
+    // to get these directly from the op
+    llvm::append_values(ptr_shape_strides, op.getBase());
+    llvm::append_range(ptr_shape_strides, castToI64(rewriter, op.getShape()));
+    llvm::append_range(ptr_shape_strides, op.getStrides());
+    rewriter.replaceOpWithMultiple(op, {ptr_shape_strides});
```

Comment:
Cool, this makes the implementation a lot easier compared to the block pointer rewrite pass.